### ISSUE TITLE
[null-safety] failedAssertion in errors_patch.dart can be null

### ISSUE THREAD

**Issue by jonahwilliams** [flutter/flutter#63513]
https://github.com/dart-lang/sdk/blob/master/sdk/lib/_internal/vm/lib/errors_patch.dart#L26

I'm unable to enable null-assertions as the act of throwing an assertions seems to be tripping the null-assertion check:

```
'dart:core-patch/errors_patch.dart': Failed assertion: line 29 pos 12: '_failedAssertion != null': is not true.
#0      _AssertionError._doThrowNewSource (dart:core-patch/errors_patch.dart:48:45)
#1      _AssertionError._throwNewNullAssertion (dart:core-patch/errors_patch.dart:42:5)
#2      new _AssertionError._create (dart:core-patch/errors_patch.dart:29:12)
#3      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:46:39)
#4      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:36:5)
#5      RawKeyboard._handleKeyEvent (package:flutter/src/services/raw_keyboard.dart:584:12)
#6      BasicMessageChannel.setMessageHandler.<anonymous closure> (package:flutter/src/services/platform_channel.dart:74:49)
#7      _DefaultBinaryMessenger.handlePlatformMessage (package:flutter/src/services/binding.dart:283:33)
#8      TestDefaultBinaryMessenger.handlePlatformMessage (package:flutter_test/src/binding.dart:130:21)
#9      main.<anonymous closure>.<anonymous closure> (file:///Users/jonahwilliams/Documents/flutter/packages/flutter/test/services/raw_keyboard_test.dart:290:14)
#10     testWidgets.<anonymous closure>.<anonymous closure> (package:flutter_test/src/widget_tester.dart:146:29)
<asynchronous suspension>

```

The assertion in question is  https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/services/raw_keyboard.dart#L584

**Comment by jonahwilliams** [comment]
This wasn't happening the last time I tried to enable null assertions, so it might be a recent change - or perhaps the assert changed in a way that is tickling this.

**Comment by alexmarkov** [comment]
@jonahwilliams How to reproduce this error? Which mode is used to compile and run this?

It looks like `_AssertionError._create` constructor expects non-null `_failedAssertion` argument.
`_failedAssertion` is supposed to be a snippet of source code for the assertion condition.
It is extracted here:
https://github.com/dart-lang/sdk/blob/552a8b1a0cc0acb8b924230533ff81d4b08987c9/runtime/lib/errors.cc#L88-L96
Maybe VM was not able to extract that source code (if sources are not available). Or maybe token positions for the assertion are not valid.

**Comment by jonahwilliams** [comment]
This can be reproduced by running `flutter test --null-assertions test/services/raw_keyboard_test.dart` on ToT. The test expectation will hide the original error a bit, some prints will help:

```
diff --git a/packages/flutter/test/services/raw_keyboard_test.dart b/packages/flutter/test/services/raw_keyboard_test.dart
index c822ff7e89..b5499b178e 100644
--- a/packages/flutter/test/services/raw_keyboard_test.dart
+++ b/packages/flutter/test/services/raw_keyboard_test.dart
@@ -303,6 +303,8 @@ void main() {
       } finally {
         FlutterError.onError = oldHandler;
       }
+      print(errorDetails.exceptionAsString());
+      print(errorDetails.stack);
       expect(errorDetails, isNotNull);
       expect(errorDetails.stack, isNotNull);
       final String fullErrorMessage = errorDetails.toString().replaceAll('\n', ' ');
```

**Comment by alexmarkov** [comment]
Could you print assertion error which is thrown from https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/services/raw_keyboard.dart#L584 without `--null-assertions`?
Does it include assertion text?

**Comment by jonahwilliams** [comment]
The assertion without null-assertions is:

```
Attempted to send a key down event when no keys are in keysPressed. This state can occur if the key event being sent doesn't properly set its modifier flags. This was the event: RawKeyDownEvent#000b3(logicalKey: LogicalKeyboardKey#700e1(keyId: "0x1000700e1", keyLabel: null, debugName: "Shift Left"), physicalKey: PhysicalKeyboardKey#700e1(usbHidUsage: "0x000700e1", debugName: "Shift Left")) and its data: RawKeyEventDataAndroid(keyLabel: null flags: 0, codePoint: 0, keyCode: 59, scanCode: 42, metaState: 0, modifiers down: {})
'package:flutter/src/services/raw_keyboard.dart':
Failed assertion: line 586 pos 12: 'null'
#0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:46:39)
#1      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:36:5)
#2      RawKeyboard._handleKeyEvent (package:flutter/src/services/raw_keyboard.dart:586:12)
#3      BasicMessageChannel.setMessageHandler.<anonymous closure> (package:flutter/src/services/platform_channel.dart:74:49)
#4      _DefaultBinaryMessenger.handlePlatformMessage (package:flutter/src/services/binding.dart:283:33)
#5      TestDefaultBinaryMessenger.handlePlatformMessage (package:flutter_test/src/binding.dart:130:21)
#6      main.<anonymous closure>.<anonymous closure> (file:///Users/jonahwilliams/Documents/flutter/packages/flutter/test/services/raw_keyboard_test.dart:290:14)
#7      testWidgets.<anonymous closure>.<anonymous closure> (package:flutter_test/src/widget_tester.dart:146:29)
<asynchronous suspension>
#8      testWidgets.<anonymous closure>.<anonymous closure> (package:flutter_test/src/widget_tester.dart)
#9      TestWidgetsFlutterBinding._runTestBody (package:flutter_test/src/binding.dart:784:19)
<asynchronous suspension>
#10     _rootRunBinary (dart:async/zone.dart:1222:13)
#11     _CustomZone.runBinary (dart:async/zone.dart:1107:19)
#12     TestWidgetsFlutterBinding._runTest (package:flutter_test/src/binding.dart:764:14)
#13     AutomatedTestWidgetsFlutterBinding.runTest.<anonymous closure> (package:flutter_test/src/binding.dart:1173:24)
#14     FakeAsync.run.<anonymous closure>.<anonymous closure> (package:fake_async/fake_async.dart:178:54)
#15     _rootRun (dart:async/zone.dart:1190:13)
#16     _CustomZone.run (dart:async/zone.dart:1093:19)
#17     _runZoned (dart:async/zone.dart:1630:10)
#18     runZoned (dart:async/zone.dart:1550:10)
#19     withClock (package:clock/src/default.dart:48:10)
#20     FakeAsync.run.<anonymous closure> (package:fake_async/fake_async.dart:178:22)
#21     _rootRun (dart:async/zone.dart:1190:13)
#22     _CustomZone.run (dart:async/zone.dart:1093:19)
#23     _runZoned (dart:async/zone.dart:1630:10)
#24     runZoned (dart:async/zone.dart:1550:10)
#25     FakeAsync.run (package:fake_async/fake_async.dart:178:7)
#26     AutomatedTestWidgetsFlutterBinding.runTest (package:flutter_test/src/binding.dart:1170:15)
#27     testWidgets.<anonymous closure> (package:flutter_test/src/widget_tester.dart:138:24)
#28     Declarer.test.<anonymous closure>.<anonymous closure> (package:test_api/src/backend/declarer.dart:175:19)
<asynchronous suspension>
#29     Declarer.test.<anonymous closure>.<anonymous closure> (package:test_api/src/backend/declarer.dart)
#30     _rootRun (dart:async/zone.dart:1190:13)
#31     _CustomZone.run (dart:async/zone.dart:1093:19)
#32     _runZoned (dart:async/zone.dart:1630:10)
#33     runZoned (dart:async/zone.dart:1550:10)
#34     Declarer.test.<anonymous closure> (package:test_api/src/backend/declarer.dart:173:13)
#35     Invoker.waitForOutstandingCallbacks.<anonymous closure> (package:test_api/src/backend/invoker.dart:231:15)
#36     _rootRun (dart:async/zone.dart:1190:13)
#37     _CustomZone.run (dart:async/zone.dart:1093:19)
#38     _runZoned (dart:async/zone.dart:1630:10)
#39     runZoned (dart:async/zone.dart:1550:10)
#40     Invoker.waitForOutstandingCallbacks (package:test_api/src/backend/invoker.dart:228:5)
#41     Invoker._onRun.<anonymous closure>.<anonymous closure>.<anonymous closure> (package:test_api/src/backend/invoker.dart:383:17)
<asynchronous suspension>
#42     Invoker._onRun.<anonymous closure>.<anonymous closure>.<anonymous closure> (package:test_api/src/backend/invoker.dart)
#43     _rootRun (dart:async/zone.dart:1190:13)
#44     _CustomZone.run (dart:async/zone.dart:1093:19)
#45     _runZoned (dart:async/zone.dart:1630:10)
#46     runZoned (dart:async/zone.dart:1550:10)
#47     Invoker._onRun.<anonymous closure>.<anonymous closure> (package:test_api/src/backend/invoker.dart:370:9)
#48     Invoker._guardIfGuarded (package:test_api/src/backend/invoker.dart:415:15)
#49     Invoker._onRun.<anonymous closure> (package:test_api/src/backend/invoker.dart:369:7)
#50     Chain.capture.<anonymous closure> (package:stack_trace/src/chain.dart:104:24)
#51     _rootRun (dart:async/zone.dart:1190:13)
#52     _CustomZone.run (dart:async/zone.dart:1093:19)
#53     _runZoned (dart:async/zone.dart:1630:10)
#54     runZoned (dart:async/zone.dart:1550:10)
#55     Chain.capture (package:stack_trace/src/chain.dart:102:12)
#56     Invoker._onRun (package:test_api/src/backend/invoker.dart:368:11)
#57     LiveTestController.run (package:test_api/src/backend/live_test_controller.dart:153:11)
#58     RemoteListener._runLiveTest.<anonymous closure> (package:test_api/src/remote_listener.dart:256:16)
#59     _rootRun (dart:async/zone.dart:1190:13)
#60     _CustomZone.run (dart:async/zone.dart:1093:19)
#61     _runZoned (dart:async/zone.dart:1630:10)
#62     runZoned (dart:async/zone.dart:1550:10)
#63     RemoteListener._runLiveTest (package:test_api/src/remote_listener.dart:255:5)
#64     RemoteListener._serializeTest.<anonymous closure> (package:test_api/src/remote_listener.dart:208:7)
#65     _rootRunUnary (dart:async/zone.dart:1198:47)
#66     _CustomZone.runUnary (dart:async/zone.dart:1100:19)
#67     _CustomZone.runUnaryGuarded (dart:async/zone.dart:1005:7)
#68     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:357:11)
#69     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:285:7)
#70     _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:808:19)
#71     _StreamController._add (dart:async/stream_controller.dart:682:7)
#72     _StreamController.add (dart:async/stream_controller.dart:624:5)
#73     _rootRunUnary (dart:async/zone.dart:1206:13)
#74     _CustomZone.runUnary (dart:async/zone.dart:1100:19)
#75     _CustomZone.runUnaryGuarded (dart:async/zone.dart:1005:7)
#76     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:357:11)
#77     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:285:7)
#78     _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:808:19)
#79     _StreamController._add (dart:async/stream_controller.dart:682:7)
#80     _StreamController.add (dart:async/stream_controller.dart:624:5)
#81     _StreamSinkWrapper.add (dart:async/stream_controller.dart:900:13)
#82     _GuaranteeSink.add (package:stream_channel/src/guarantee_channel.dart:125:12)
#83     new _MultiChannel.<anonymous closure> (package:stream_channel/src/multi_channel.dart:159:31)
#84     _rootRunUnary (dart:async/zone.dart:1198:47)
#85     _CustomZone.runUnary (dart:async/zone.dart:1100:19)
#86     _CustomZone.runUnaryGuarded (dart:async/zone.dart:1005:7)
#87     CastStreamSubscription._onData (dart:_internal/async_cast.dart:85:11)
#88     _rootRunUnary (dart:async/zone.dart:1198:47)
#89     _CustomZone.runUnary (dart:async/zone.dart:1100:19)
#90     _CustomZone.runUnaryGuarded (dart:async/zone.dart:1005:7)
#91     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:357:11)
#92     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:285:7)
#93     _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:808:19)
#94     _StreamController._add (dart:async/stream_controller.dart:682:7)
#95     _StreamController.add (dart:async/stream_controller.dart:624:5)
#96     _rootRunUnary (dart:async/zone.dart:1198:47)
#97     _CustomZone.runUnary (dart:async/zone.dart:1100:19)
#98     _CustomZone.runUnaryGuarded (dart:async/zone.dart:1005:7)
#99     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:357:11)
#100    _BufferingStreamSubscription._add (dart:async/stream_impl.dart:285:7)
#101    _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:808:19)
#102    _StreamController._add (dart:async/stream_controller.dart:682:7)
#103    _StreamController.add (dart:async/stream_controller.dart:624:5)
#104    _StreamSinkWrapper.add (dart:async/stream_controller.dart:900:13)
#105    _rootRunUnary (dart:async/zone.dart:1198:47)
#106    _CustomZone.runUnary (dart:async/zone.dart:1100:19)
#107    _CustomZone.runUnaryGuarded (dart:async/zone.dart:1005:7)
#108    _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:357:11)
#109    _BufferingStreamSubscription._add (dart:async/stream_impl.dart:285:7)
#110    _ForwardingStreamSubscription._add (dart:async/stream_pipe.dart:127:11)
#111    _MapStream._handleData (dart:async/stream_pipe.dart:224:10)
#112    _ForwardingStreamSubscription._handleData (dart:async/stream_pipe.dart:157:13)
#113    _rootRunUnary (dart:async/zone.dart:1198:47)
#114    _CustomZone.runUnary (dart:async/zone.dart:1100:19)
#115    _CustomZone.runUnaryGuarded (dart:async/zone.dart:1005:7)
#116    _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:357:11)
#117    _BufferingStreamSubscription._add (dart:async/stream_impl.dart:285:7)
#118    _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:808:19)
#119    _StreamController._add (dart:async/stream_controller.dart:682:7)
#120    _StreamController.add (dart:async/stream_controller.dart:624:5)
#121    new _WebSocketImpl._fromSocket.<anonymous closure> (dart:_http/websocket_impl.dart:1145:21)
#122    _rootRunUnary (dart:async/zone.dart:1198:47)
#123    _CustomZone.runUnary (dart:async/zone.dart:1100:19)
#124    _CustomZone.runUnaryGuarded (dart:async/zone.dart:1005:7)
#125    _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:357:11)
#126    _BufferingStreamSubscription._add (dart:async/stream_impl.dart:285:7)
#127    _SinkTransformerStreamSubscription._add (dart:async/stream_transformers.dart:69:11)
#128    _EventSinkWrapper.add (dart:async/stream_transformers.dart:15:11)
#129    _WebSocketProtocolTransformer._messageFrameEnd (dart:_http/websocket_impl.dart:338:23)
#130    _WebSocketProtocolTransformer.add (dart:_http/websocket_impl.dart:232:46)
#131    _SinkTransformerStreamSubscription._handleData (dart:async/stream_transformers.dart:121:24)
#132    _rootRunUnary (dart:async/zone.dart:1198:47)
#133    _CustomZone.runUnary (dart:async/zone.dart:1100:19)
#134    _CustomZone.runUnaryGuarded (dart:async/zone.dart:1005:7)
#135    _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:357:11)
#136    _BufferingStreamSubscription._add (dart:async/stream_impl.dart:285:7)
#137    _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:808:19)
#138    _StreamController._add (dart:async/stream_controller.dart:682:7)
#139    _StreamController.add (dart:async/stream_controller.dart:624:5)
#140    _Socket._onData (dart:io-patch/socket_patch.dart:2043:41)
#141    _rootRunUnary (dart:async/zone.dart:1206:13)
#142    _CustomZone.runUnary (dart:async/zone.dart:1100:19)
#143    _CustomZone.runUnaryGuarded (dart:async/zone.dart:1005:7)
#144    _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:357:11)
#145    _BufferingStreamSubscription._add (dart:async/stream_impl.dart:285:7)
#146    _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:808:19)
#147    _StreamController._add (dart:async/stream_controller.dart:682:7)
#148    _StreamController.add (dart:async/stream_controller.dart:624:5)
#149    new _RawSocket.<anonymous closure> (dart:io-patch/socket_patch.dart:1579:33)
#150    _NativeSocket.issueReadEvent.issue (dart:io-patch/socket_patch.dart:1075:14)
#151    _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
#152    _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
```

`Failed assertion: line 586 pos 12: 'null'` seems suspicious, since this assertion is not a null check - is this where the sources are getting dropped?

**Comment by leafpetersen** [comment]
It looks to me like `_failedAssertion` is being passed in as `null`, and the type of `_failedAssertion` doesn't accept `null`.  So either the type of `_failedAssertion` was migrated wrong, or whatever code path is sending `null` through needs to be fixed.  Or is the claim that `_failedAssertion` is not null but the assertion is still firing?

**Comment by alexmarkov** [comment]
@jonahwilliams Yes, it seems like VM failed to get a source text for raw_keyboard.dart.
Do you know how `flutter test` compiles and runs Dart code? Does it run from kernel files? If yes, could you double check if those kernel files contain source text. In front-end server source text can be included with `--embed-source-text` option.

VM should not really try to pass null to _AssertionError constructor, this needs to be fixed. We can pass something like `<source is not available>` in case source text is not available. Ideally we should just fix https://github.com/dart-lang/sdk/issues/34586.

**Comment by alexmarkov** [comment]
Fix for failing null assertion: https://dart-review.googlesource.com/c/sdk/+/158162

**Comment by jonahwilliams** [comment]
Thank you @alexmarkov !

**Comment by alexmarkov** [comment]
The fix was rolled into Flutter in b2c0970aa7b5334580c627316e2a07500c7d9c64.
